File: benchmark/04_quicksort/quicksort.go
// Benchmark 04: QuickSort
// Category: Discrete Mathematics / Sorting
// Tests: Recursion, array manipulation, comparisons

package main

import (
	"fmt"
	"time"
)

const SIZE = 100000 // 100k elements

func swap(arr []int, i, j int) {
	arr[i], arr[j] = arr[j], arr[i]
}

func partition(arr []int, low, high int) int {
	pivot := arr[high]
	i := low - 1

	for j := low; j < high; j++ {
		if arr[j] <= pivot {
			i++
			swap(arr, i, j)
		}
	}
	swap(arr, i+1, high)
	return i + 1
}

func quickSort(arr []int, low, high int) {
	if low < high {
		pi := partition(arr, low, high)
		quickSort(arr, low, pi-1)
		quickSort(arr, pi+1, high)
	}
}

func main() {
	start := time.Now()

	// Allocate and initialize array
	arr := make([]int, SIZE)
	for i := 0; i < SIZE; i++ {
		arr[i] = (SIZE - i) * 17 % SIZE
	}

	// Sort
	quickSort(arr, 0, SIZE-1)

	elapsed := time.Since(start)

	// Verify sorted
	sorted := true
	for i := 1; i < SIZE; i++ {
		if arr[i] < arr[i-1] {
			sorted = false
			break
		}
	}

	fmt.Printf("Array size: %d\n", SIZE)
	fmt.Printf("Sorted correctly: %v\n", sorted)
	fmt.Printf("Time: %.4f seconds\n", elapsed.Seconds())
}

File: benchmark/04_quicksort/quicksort.rs
// Benchmark 04: QuickSort
// Category: Discrete Mathematics / Sorting
// Tests: Recursion, array manipulation, comparisons

use std::time::Instant;

const SIZE: usize = 100_000; // 100k elements

fn partition(arr: &mut [i32], low: usize, high: usize) -> usize {
    let pivot = arr[high];
    let mut i = low as i32 - 1;

    for j in low..high {
        if arr[j] <= pivot {
            i += 1;
            arr.swap(i as usize, j);
        }
    }
    arr.swap((i + 1) as usize, high);
    (i + 1) as usize
}

fn quick_sort(arr: &mut [i32], low: usize, high: usize) {
    if low < high {
        let pi = partition(arr, low, high);
        if pi > 0 {
            quick_sort(arr, low, pi - 1);
        }
        quick_sort(arr, pi + 1, high);
    }
}

fn main() {
    let start = Instant::now();

    // Allocate and initialize array
    let mut arr: Vec<i32> = (0..SIZE).map(|i| ((SIZE - i) * 17 % SIZE) as i32).collect();

    // Sort
    quick_sort(&mut arr, 0, SIZE - 1);

    let elapsed = start.elapsed();

    // Verify sorted
    let sorted = arr.windows(2).all(|w| w[0] <= w[1]);

    println!("Array size: {}", SIZE);
    println!("Sorted correctly: {}", sorted);
    println!("Time: {:.4} seconds", elapsed.as_secs_f64());
}
